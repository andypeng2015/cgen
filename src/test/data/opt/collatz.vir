module foo

export function w $foo(w %n) {
  %count = slot 4, align 4
  %x = slot 4, align 4
@start:
  st.w 0_w, %count 
  st.w %n, %x
  jmp @header
@header:
  %a = ld.w %x
  %c = gt.w %a, 1_w
  br %c, @body, @done
@body:
  %a = ld.w %x
  %r = urem.w %a, 2_w
  %c = eq.w %r, 0_w
  br %c, @eqz, @neqz
@eqz:
  %a = ld.w %x
  %d = udiv.w %a, 2_w
  st.w %d, %x
  jmp @inc
@neqz:
  %a = ld.w %x
  %m = mul.w %a, 3_w
  %m = add.w %m, 1_w
  st.w %m, %x
  jmp @inc
@inc:
  %k = ld.w %count
  %k = add.w %k, 1_w
  st.w %k, %count
  jmp @header
@done:
  %k = ld.w %count
  ret %k
}
