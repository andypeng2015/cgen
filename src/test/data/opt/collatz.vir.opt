module foo

export function w $foo(w %n) {
@2:
  jmp @3(%n, 0x0_w)
@3(%x.2, %count.2):
  %8 = gt.w %x.2, 0x1_w ; @35
  br %8, @6, @7
@6:
  %9 = and.w %x.2, 0x1_w ; @36
  %10 = eq.w %9, 0x0_w ; @37
  br %10, @10, @11
@10:
  %15 = lsr.w %x.2, 0x1_w ; @42
  jmp @15(%15)
@11:
  %12 = lsl.w %x.2, 0x1_w ; @39
  %13 = add.w %12, %x.2 ; @40
  %14 = add.w %13, 0x1_w ; @41
  jmp @15(%14)
@15(%x.3):
  %11 = add.w %count.2, 0x1_w ; @38
  jmp @3(%x.3, %11)
@7:
  ret %count.2
}
