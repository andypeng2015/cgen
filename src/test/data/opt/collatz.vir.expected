module foo

export function w $foo(w %n) {
@0:
  jmp @1(%n, 0x0_w)
@1(%x.2, %count.2):
  %0 = gt.w %x.2, 0x1_w ; @25
  br %0, @4, @5
@4:
  %1 = and.w %x.2, 0x1_w ; @26
  %2 = eq.w %1, 0x0_w ; @27
  br %2, @8, @9
@8:
  %7 = lsr.w %x.2, 0x1_w ; @32
  jmp @13(%7)
@9:
  %4 = lsl.w %x.2, 0x1_w ; @29
  %5 = add.w %4, %x.2 ; @30
  %6 = add.w %5, 0x1_w ; @31
  jmp @13(%6)
@13(%x.3):
  %3 = add.w %count.2, 0x1_w ; @28
  jmp @1(%x.3, %3)
@5:
  ret %count.2
}
